### ISSUE TITLE
Run syncNamespaces before syncPods

### ISSUE THREAD

**Issue by girishmg** [ovn-org/ovn-kubernetes#1418]
The OVN Master controller upon start does this:

```
func (oc *Controller) Run() error {
	oc.syncPeriodic()
	// WatchNodes must be started first so that its initial Add will
	// create all node logical switches, which other watches may depend on.
	// https://github.com/ovn-org/ovn-kubernetes/pull/859
	if err := oc.WatchNodes(); err != nil {
		return err
	}

	for _, f := range []func() error{oc.WatchPods, oc.WatchServices, oc.WatchEndpoints,
		oc.WatchNamespaces, oc.WatchNetworkPolicy} {
		if err := f(); err != nil {
			return err
		}
	}
}
```

So, it calls each of these watchers on the resources it cares about. Each of these watchers in turn call onAdd() handler for each of the corresponding existing resource objects. These calls are all synchronous.

I think the WatchNamespaces should be called immediately after the WatchNodes, but before WatchPods. The WatchNamespaces sets up some scaffolding (AddressSet, for example) for the existing namespaces, which lot of other resources like pod, service, networkPolicy, and so on depend on. Right now we have code in certain resources, which gets called before the namespace is setup, that waits on Namespace resources to be setup before it can proceed. All this code can be torched or at the very least these code will return soon since the namespace resource is already setup.

With the current order we have today, I also found a potential damage that we could cause in the code below

```
// AddNamespace creates corresponding addressset in ovn db
func (oc *Controller) AddNamespace(ns *kapi.Namespace) {
	klog.V(5).Infof("Adding namespace: %s", ns.Name)
	nsInfo := oc.createNamespaceLocked(ns.Name)
	defer nsInfo.Unlock()

	// Get all the pods in the namespace and append their IP to the
	// address_set
	var ips []net.IP
	existingPods, err := oc.watchFactory.GetPods(ns.Name)
	if err != nil {
		klog.Errorf("Failed to get all the pods (%v)", err)
	} else {
        }
}
```

Say, in the above function we failed in GetPods(). This translate to `nil` ips. Which translates into us clearing the address_set in call to newOvnAddressSet(). It is hard to differentiate ips being `nil` because there were no pods in the namespace or GetPods() failed inside the newOvnAddressSet() function. Now, the address_set entry for this namespace is empty even though there are pods in it.

If we call WatchNamespaces() early, then GetPods() will always return `nil`, then we clear the address_set. Then when WatchPods() gets called, the forEachPod(onAdd) will fill the NameSpace's addressSet with the Pod's IP one-by-one. 

Any thoughts on this?

**Comment by dave-tucker** [comment]
I'm just about to put up a PR to completely change/remove all of this as part of my quest to move to workqueues :)

The plan is to start all of the informers at the same time when we call to `SharedIndexInformerFactory.Start()` at this point the informers will then start populating their workqueues.

We then start all of the our `informer.EventHandler` async worker threads which will start processing these Add/Delete events in the workqueues.

The nice thing about workqueues is that we can re-queue events if we encounter an error, for example when something that we expect to exist (i.e a logical switch) doesn't exist yet.
The re-queue is rate-limited and the default is, IIRC, exponential backoff.

So I guess what I'm saying is that hopefully when we move to a more async startup that this shouldn't be an issue.

**Comment by girishmg** [comment]
@dave-tucker I am looking forward for your PR. Any write up you may have, please share it so that we are aware as to what to expect.

For example: if we re-queue a failed event handler, say NodeAdd, then will it get queued behind the NodeUpdate for the same node? Then we will miss Node's update.  Thanks.

**Comment by dave-tucker** [comment]
Sure I can write something up and send it to the mailing list.

**Comment by girishmg** [comment]
@dave-tucker looks like @trozet merged this PR #1438 that changes the order as captured in this issue.
